alf/algorithms/misc_algorithm.py

```diff
@@ -0,0 +1,239 @@
+                hidden_size, input_shape=soi_spec.shape, activation=None)
+            misc_layero = tf.keras.layers.Dense(
+                1, input_shape=(None, None, hidden_size), activation=None)
+
+        self._misc_layerx = misc_layerx
+        self._misc_layery = misc_layery
+        self._misc_layero = misc_layero
+
+        self._traj_spec = traj_spec
+        self._buffer_size = buffer_size
+        self._buffer = DataBuffer(self._traj_spec, capacity=self._buffer_size)
+        self.mi_r_scale = mi_r_scale
+        self.n_objects = n_objects
+        self.empowerment = empowerment
+
+    def split_observation_tf(self, o, env_name='SocialBot-PlayGround-v0'):
+
+        dimo = o.get_shape().as_list()[-1]
+
+        if env_name in ['SocialBot-PlayGround-v0']:
+            if dimo == 21:
+                task_specific_ob, agent_pose, agent_vel, internal_states, action = tf.split(
+                    o, [3, 6, 6, 4, 2], axis=-1)
+
+                agent_pose_1, agent_pose_2 = tf.split(
+                    agent_pose, [3, 3], axis=-1)
+                joint_pose_1, joint_pose_2, joint_vel_1, joint_vel_2 = tf.split(
+                    internal_states, [1, 1, 1, 1], axis=-1)
+                joint_1 = tf.concat([joint_pose_1, joint_vel_1], axis=-1)
+                joint_2 = tf.concat([joint_pose_2, joint_vel_2], axis=-1)
+
+                if self.n_objects == 1:
+                    obs_achieved_goal = task_specific_ob
+                    obs_excludes_goal = agent_pose_1
+                    if self.empowerment:
+                        obs_excludes_goal = action
+                elif self.n_objects == 0:
+                    obs_achieved_goal = joint_1
+                    obs_excludes_goal = joint_2
+
+                return (obs_excludes_goal, obs_achieved_goal)
+
+            elif (dimo == 24) and (self.n_objects == 2):
+                task_specific_ob_1, task_specific_ob_2, agent_pose, agent_vel, internal_states, action = tf.split(
+                    o, [3, 3, 6, 6, 4, 2], axis=-1)
+
+                agent_pose_1, agent_pose_2 = tf.split(
+                    agent_pose, [3, 3], axis=-1)
+
+                obs_achieved_goal_1 = task_specific_ob_1
+                obs_achieved_goal_2 = task_specific_ob_2
+                obs_excludes_goal = agent_pose_1
+
+                return (obs_excludes_goal, obs_achieved_goal_1,
+                        obs_achieved_goal_2)
+            else:
+                print(
+                    'Please specify the state of interests and the context state in the misc_alorighm.py first.'
+                )
+                exit()
+        else:
+            print(
+                'Please specify the state of interests and the context state in the misc_alorighm.py first.'
+            )
+            exit()
+
+    def mine(self, x_in, y_in):
+
+        # Mutual Information Neural Estimation
+        y_in_tran = tf.transpose(y_in, perm=[1, 0, 2])
+        y_shuffle_tran = tf.gather(
```

Comment:
add a comment to explain why not use tr.random.shuffle(y_in_tran)

Reply:
added